packages/go_router/lib/src/match.dart

```diff
@@ -263,8 +263,11 @@ class RouteMatchList {
     assert(index != -1);
     newMatches.removeRange(index, newMatches.length);
 
-    // Also pop ShellRoutes when there are no subsequent route matches
-    while (newMatches.isNotEmpty && newMatches.last.route is ShellRouteBase) {
+    // Also pop ShellRoutes that have no subsequent route matches and GoRoutes
+    // that only have redirect.
+    while (newMatches.isNotEmpty &&
+        (newMatches.last.route is ShellRouteBase ||
+            (newMatches.last.route as GoRoute).redirectOnly)) {
```

Comment:
Consider using Dart's exhaustiveness to check the type, so that we don't miss any cases:

```dart
while(newMatches.isNotEmpty) {
  var last = newMatches.last.route;
  switch(last) {
    case ShellRouteBase():
      newMatches.removeLast();
    case GoRoute(redirectOnly: true):
      newMatches.removeLast();
    case _:
      // ??
  }
}

Reply:
I don't think flutter package support dart 3 as minimum version yet?

Reply:
Ah, okay.